docs/design/metainfo-rpc.md

```diff
@@ -26,87 +26,211 @@ The following is a proto file for just the rpc portion to create segments. The r
 With this design, streams will be created that contain multiple parts. There were already RPCs to finalize individual segments, but each segment was not logically associated with a part, but instead just the specific object.
 
 ```protobuf
-message AddressedOrderLimit {
-    orders.OrderLimit2 limit = 1;
-    node.NodeAddress storage_node_address = 2;
+service Metainfo {
+    rpc CreateBucket(BucketCreateRequest) returns (BucketCreateResponse);
+    rpc GetBucket(BucketGetRequest) returns (BucketGetResponse);
+    rpc DeleteBucket(BucketDeleteRequest) returns (BucketDeleteResponse);
+    rpc ListBuckets(BucketListRequest) returns (BucketListResponse);
+
+    rpc CreateObject(ObjectCreateRequest) returns (ObjectCreateResponse);
```

Comment:
Since I can't imagine wanting to create an object without also creating the first segment, and since creating the segment requires information from creating the object (the stream_id), there really should be a combined RPC for doing both to reduce a RTT.

Reply:
Additionally, there should be a combined RPC for committing a segment and an object because you can't async send both: you don't know if the server will receive the object commit before the segment commit and have that segment not be included. That reduces a RTT for the commit as well.

Reply:
just wondering, why are multiple computers uploading different segments as part of one object? is that a normal thing? also if they're doing that, wouldn't the combined rpc still be good to ensure the create object happens before the create segment?

Reply:
It's important to have the common case be as fast as possible, and the common case is uploading an object from a single computer with possibly a single segment. Not having combined RPCs penalizes the common case to 2x the RTTs as it with combined RPCs.

Reply:
https://grpc.io/docs/tutorials/async/helloasync-cpp/ for RTT issues. I'm not quite sure how difficult it would be to support on Go. If that's really complicated I'm fine with adding the specific versions, but we still need the separate versions for concurrent upload.

Reply:
I realized one thing we could do is have a message for batch requests:

```
message BatchRequest {
    message Request {
        ObjectCreateRequest object_create;
        ObjectCommitRequest object_commit;
    
        SegmentCreateRequest segment_create;
        SegmentCommitRequest segment_commit;
        SegmentInlineRequest segment_inline;
    }
    repeated Request requests;
}
```

This would allow to implement:
```
object create
segment inline
object commit
```

In a single roundtrip. Also we can enforce the semantics to be such that the requests will be executed until first of them fails, which is similar to invoking them individually, but getting the RTT improvements.

Reply:
Added BatchRequest/BatchResponse

Reply:
> ...we still need the separate versions for concurrent upload

I'm trying to visualize a use case that requires having a version for an Object.
Would you mind to write a brief one here?

Reply:
3 concurrent uploads on the same encrypted path, one of them fails and one of them is canceled.

Reply:
so for avoiding to override the data on the same path without being aware?

Reply:
I think this solves all of my RTT concerns.